flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperatorStateHandler.java

```diff
@@ -0,0 +1,434 @@
+
+			if (null != operatorStateBackend) {
+				snapshotInProgress.setOperatorStateManagedFuture(
+					operatorStateBackend.snapshot(checkpointId, timestamp, factory, checkpointOptions));
+			}
+
+			if (null != keyedStateBackend) {
+				snapshotInProgress.setKeyedStateManagedFuture(
+					keyedStateBackend.snapshot(checkpointId, timestamp, factory, checkpointOptions));
+			}
+		} catch (Exception snapshotException) {
+			try {
+				snapshotInProgress.cancel();
+			} catch (Exception e) {
+				snapshotException.addSuppressed(e);
+			}
+
+			String snapshotFailMessage = "Could not complete snapshot " + checkpointId + " for operator " +
+				operatorName + ".";
+
+			try {
+				snapshotContext.closeExceptionally();
+			} catch (IOException e) {
+				snapshotException.addSuppressed(e);
+			}
+			throw new CheckpointException(snapshotFailMessage, CheckpointFailureReason.CHECKPOINT_DECLINED, snapshotException);
+		}
+
+		return snapshotInProgress;
+	}
+
+	/**
+	 * Stream operators with state, which want to participate in a snapshot need to override this hook method.
+	 *
+	 * @param context context that provides information and means required for taking a snapshot
+	 * @param operatorName
+	 */
+	public void snapshotState(StateSnapshotContext context, String operatorName) throws Exception {
+		final KeyedStateBackend<?> keyedStateBackend = getKeyedStateBackend();
+		//TODO all of this can be removed once heap-based timers are integrated with RocksDB incremental snapshots
+		if (keyedStateBackend instanceof AbstractKeyedStateBackend &&
+			((AbstractKeyedStateBackend<?>) keyedStateBackend).requiresLegacySynchronousTimerSnapshots()) {
+
+			KeyedStateCheckpointOutputStream out;
+
+			try {
+				out = context.getRawKeyedOperatorStateOutput();
+			} catch (Exception exception) {
+				throw new Exception("Could not open raw keyed operator state stream for " +
+					operatorName + '.', exception);
+			}
+
+			try {
+				KeyGroupsList allKeyGroups = out.getKeyGroupList();
+				for (int keyGroupIdx : allKeyGroups) {
+					out.startNewKeyGroup(keyGroupIdx);
+
+					timeServiceManager.snapshotStateForKeyGroup(
+						new DataOutputViewStreamWrapper(out), keyGroupIdx);
+				}
+			} catch (Exception exception) {
+				throw new Exception("Could not write timer service of " + operatorName +
+					" to checkpoint state stream.", exception);
+			} finally {
+				try {
+					out.close();
+				} catch (Exception closeException) {
+					LOG.warn("Could not close raw keyed operator state stream for {}. This " +
+						"might have prevented deleting some state data.", operatorName, closeException);
+				}
+			}
+		}
+	}
+
+	public void notifyCheckpointComplete(long checkpointId) throws Exception {
+		if (keyedStateBackend != null) {
+			keyedStateBackend.notifyCheckpointComplete(checkpointId);
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	public <K> KeyedStateBackend<K> getKeyedStateBackend() {
+		return (KeyedStateBackend<K>) keyedStateBackend;
+	}
```

Comment:
1. I don't think that *abstract* class should have any public API. Ideally, only it's descendants should know about it.
2. It sounds very strange that `keyedStateBackend` is used in not keyed context. Do you mean non-keyed operators? Should we split it then into non-keyed and keyed versions?
If not, this class looks keyed and most of its clients know their keys; those who don't can use wildcards/`<Object>`/casts/....or be fixed:)

Reply:
1. this is not an abstract class, and I don't see a problem of exposing public methods in an abstract class.
2. 
>  Do you mean non-keyed operators

yes. This is one of the reasons why I was struggling with figuring out how the operator API should look like.

>  Should we split it then into non-keyed and keyed versions?

I don't know. As I wrote in my [overall comment](https://github.com/apache/flink/pull/11403#issuecomment-600149707):

> If you would like to split it even further to things like keyed and not-keyed operators, those two problems (spaghetti/web of connections and multiple inheritance) are going to be more profound.
>
> I don't know at the moment how to tackle those problems, but for me they are out of scope of this PR/JIRA/FLIP :/

Reply:
1. I don't like fanatic approach to restricting visibility, especially if it doesn't make sense. For me those methods are valid public APIs of this class, it's intention is to expose those things. Artificially hiding theirs visibility can cause issues in the future. Like if class is re-used in different package, it will require small changes and/or block people from re-using it.
2. I don't like the idea of parametrising the whole class with key type, even for non-keyed contexts.

Reply:
1. The approach of narrowing visibility is a practical one: if a method is `public` then it's part of a class API; and somebody will eventually use it; and it will be difficult to change it later. On the other hand, if and when you *actually* need it, you can widen the visibility.
2. Why, can you please elaborate?

As it is now, I see:
- `AbstractKeyedStateBackend<?> keyedStateBackend;` (<**?**>)
- casts: in `StreamOperatorStateHandler` AND in some operators like `final K key = this.<K>getKeyedStateBackend().getCurrentKey();` 
- `@SuppressWarnings` all the way to actual usages
- confusing type parameter in method  (can I pass different types in different calls? is it the same as for `getInternalTimerService`?)

Each of these places is a potential bug.

What I propose is to have
`KeyedOp<K> extends AbstractStreamOperatorV2<K>`
and 
`NonKeyedOp extends AbstractStreamOperatorV2<Object>`.
(and  `StreamOperatorStateHandler<T>` inside)

Reply:
2. `KeyedOp` and `NonKeyedOp` won't work nicely, unless you want to multiply number of abstract classes that are extending from `AbstractStreamOperator`/`AbstractStreamOperatorV2` by factor of two.

Reply:
I think we should parameterize new classes (`AbstractStreamOperatorV2` and `StreamOperatorStateHandler`) and keep the old one (`AbstractStreamOperator`) as is.

@pnowojski , I didn't get your point here:
> KeyedOp and NonKeyedOp won't work nicely, unless you want to multiply number of abstract classes that are extending from AbstractStreamOperator/AbstractStreamOperatorV2 by factor of two.

Can you please explain if it's still relevant?

Reply:
As I mentioned before, there are other classes that are building on top of `AbstractStreamOperator`, like `AbstractUdfStreamOperator`, `AbstractPythonFunctionOperator`, `TableStreamOperator`, .... With splitting `AbstractStreamOperator` into more classes like keyed, non-keyed, one input, two input, multi input, we multiply the number of classes that need to extend from it.

Reply:
I meant to add type parameter to them, not splitting them. Sorry for the confusion.
(we don't have to update the whole hierarchy while migrating; e.g. we can start with
`AbstractUdfStreamOperator<OUT> extends AbstractStreamOperatorV2<Object, OUT>`.)

Reply:
Summarizing the offline discussion about type parameters:
1. For `AbstractStreamOperatorV2`, it's unclear whether it's a public API for `MultipleInputStreamOperator`s or not. And if it is, then it's unclear whether it should reproduce `AbstractStreamOperator` API. But to be on the safe side it's better to have the same API (i.e. type parameters on method level, not on class).
1. For `StreamOperatorStateHandler`, this is a trade-off between confusion by `extends<Object>`, by method type parameters, unnecessary casts, duplication of these casts.  Given that this is reversible, it's better to go with type parameters for methods for now as well.